File: bundle-plugin/src/test/scala/com/yahoo/container/plugin/osgi/ImportPackageTest.scala
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.osgi

import org.scalatest.junit.{JUnitSuite, ShouldMatchersForJUnit}
import org.junit.Test

import ImportPackages.Import
import ExportPackages.{Parameter, Export}

/**
 * @author  tonytv
 */
class ImportPackageTest extends JUnitSuite with ShouldMatchersForJUnit {
  val referencedPackages = Set("com.yahoo.exported")
  val exports = exportByPackageName(Export(List("com.yahoo.exported"), List()))
  val exportsWithVersion = exportByPackageName(exports.head._2.copy(parameters = List(Parameter("version", "1.3"))))

  def exportByPackageName(export : Export) = ExportPackages.exportsByPackageName(List(export))
  @Test
  def require_that_non_implemented_import_with_matching_export_is_included() {
    val imports = calculateImports(referencedPackages, implementedPackages = Set(), exportedPackages = exports)
    imports should be (Set(Import("com.yahoo.exported", None)))
  }


  @Test
  def require_that_non_implemented_import_without_matching_export_is_excluded() {
    val imports = calculateImports(referencedPackages, implementedPackages = Set(), exportedPackages = Map())
    imports should be (Set())
  }

  @Test
  def require_that_implemented_import_with_matching_export_is_excluded() {
    val imports = calculateImports(
      referencedPackages,
      implementedPackages = referencedPackages,
      exportedPackages = exports)

    imports should be (Set())
  }

  @Test
  def require_that_version_is_included() {
    val imports = calculateImports(referencedPackages, implementedPackages = Set(), exportedPackages = exportsWithVersion)

    imports should be (Set(Import("com.yahoo.exported", Some("1.3"))))
  }

  @Test
  def require_that_all_versions_up_to_the_next_major_version_is_in_range() {
    Import("foo", Some("1.2")).importVersionRange should be (Some("[1.2,2)"))
  }

  // TODO: Detecting guava packages should be based on bundle-symbolicName, not package name.
  @Test
  def require_that_for_guava_all_future_major_versions_are_in_range() {
    val rangeWithInfiniteUpperLimit = Some("[18.1," + ImportPackages.InfiniteVersion + ")")
    Import("com.google.common",     Some("18.1")).importVersionRange should be (rangeWithInfiniteUpperLimit)
    Import("com.google.common.foo", Some("18.1")).importVersionRange should be (rangeWithInfiniteUpperLimit)

    Import("com.google.commonality", Some("18.1")).importVersionRange should be (Some("[18.1,19)"))
  }

  @Test
  def require_that_none_version_gives_non_version_range() {
    Import("foo", None).importVersionRange should be (None)
  }

  @Test
  def require_that_exception_is_thrown_when_major_component_is_non_numeric() {
    intercept[IllegalArgumentException](Import("foo", Some("1notValid.2")))
  }

  @Test
  def require_that_osgi_import_supports_missing_version() {
    Import("com.yahoo.exported", None).asOsgiImport should be ("com.yahoo.exported")
  }

  @Test
  def require_that_osgi_import_version_range_includes_all_versions_from_the_current_up_to_the_next_major_version() {
    Import("com.yahoo.exported", Some("1.2")).asOsgiImport should be ("com.yahoo.exported;version=\"[1.2,2)\"")
  }

  @Test
  def require_that_osgi_import_version_range_ignores_qualifier() {
    Import("com.yahoo.exported", Some("1.2.3.qualifier")).asOsgiImport should be ("com.yahoo.exported;version=\"[1.2.3,2)\"")
  }


  def calculateImports(referencedPackages : Set[String],
                       implementedPackages : Set[String],
                       exportedPackages : Map[String, Export]) : Set[Import] = {
    ImportPackages.calculateImports(referencedPackages, implementedPackages, exportedPackages).values.toSet
  }
}

File: bundle-plugin/src/test/java/com/yahoo/container/plugin/osgi/ImportPackageTest.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.osgi;

import com.yahoo.container.plugin.osgi.ExportPackages.Export;
import com.yahoo.container.plugin.osgi.ExportPackages.Parameter;
import com.yahoo.container.plugin.osgi.ImportPackages.Import;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * @author Tony Vaagenes
 * @author ollivir
 */
public class ImportPackageTest {
    private static final String PACKAGE_NAME = "com.yahoo.exported";
    private final Set<String> referencedPackages = Collections.singleton(PACKAGE_NAME);
    private final Map<String, Export> exports = exportByPackageName(new Export(List.of(PACKAGE_NAME), Collections.emptyList()));
    private final Map<String, Export> exportsWithVersion = exportByPackageName(
            new Export(List.of(PACKAGE_NAME), List.of(new Parameter("version", "1.3"))));

    private static Map<String, Export> exportByPackageName(Export export) {
        return ExportPackages.exportsByPackageName(List.of(export));
    }

    @Test
    void require_that_non_implemented_import_with_matching_export_is_included() {
        Set<Import> imports = calculateImports(referencedPackages, Set.of(), exports);
        assertEquals(1, imports.stream().filter(imp -> imp.packageName().equals(PACKAGE_NAME) && imp.version().isEmpty()).count());
    }

    @Test
    void require_that_non_implemented_import_without_matching_export_is_excluded() {
        Set<Import> imports = calculateImports(referencedPackages, Set.of(), Map.of());
        assertTrue(imports.isEmpty());
    }

    @Test
    void require_that_implemented_import_with_matching_export_is_excluded() {
        Set<Import> imports = calculateImports(referencedPackages, referencedPackages, exports);
        assertTrue(imports.isEmpty());
    }

    @Test
    void require_that_version_is_included() {
        Set<Import> imports = calculateImports(referencedPackages, Set.of(), exportsWithVersion);
        assertEquals(1, imports.stream().filter(imp -> imp.packageName().equals(PACKAGE_NAME) && imp.version().equals("1.3")).count());
    }

    @Test
    void require_that_all_versions_up_to_the_next_major_version_is_in_range() {
        assertEquals("[1.2,2)", new Import("foo", Optional.of("1.2")).importVersionRange().get());
    }

    // TODO: Detecting guava packages should be based on bundle-symbolicName, not package name.
    @Test
    void require_that_for_guava_all_future_major_versions_are_in_range() {
        Optional<String> rangeWithInfiniteUpperLimit = Optional.of("[18.1," + ImportPackages.INFINITE_VERSION + ")");
        assertEquals(rangeWithInfiniteUpperLimit, new Import("com.google.common", Optional.of("18.1")).importVersionRange());
        assertEquals(rangeWithInfiniteUpperLimit, new Import("com.google.common.foo", Optional.of("18.1")).importVersionRange());
        assertEquals(Optional.of("[18.1,19)"), new Import("com.google.commonality", Optional.of("18.1")).importVersionRange());
    }

    @Test
    void require_that_none_version_gives_non_version_range() {
        assertTrue(new Import("foo", Optional.empty()).importVersionRange().isEmpty());
    }

    @Test
    void require_that_exception_is_thrown_when_major_component_is_non_numeric() {
        assertThrows(IllegalArgumentException.class, () -> {
            new Import("foo", Optional.of("1notValid.2"));
        });
    }

    @Test
    void require_that_osgi_import_supports_missing_version() {
        assertEquals("com.yahoo.exported", new Import("com.yahoo.exported", Optional.empty()).asOsgiImport());
    }

    @Test
    void require_that_osgi_import_version_range_includes_all_versions_from_the_current_up_to_the_next_major_version() {
        assertEquals("com.yahoo.exported;version=\"[1.2,2)\"", new Import("com.yahoo.exported", Optional.of("1.2")).asOsgiImport());
    }

    @Test
    void require_that_osgi_import_version_range_ignores_qualifier() {
        assertEquals("com.yahoo.exported;version=\"[1.2.3,2)\"", new Import("com.yahoo.exported", Optional.of("1.2.3.qualifier")).asOsgiImport());
    }

    private static Set<Import> calculateImports(Set<String> referencedPackages,
                                                Set<String> implementedPackages,
                                                Map<String, Export> exportedPackages) {
        return new HashSet<>(ImportPackages.calculateImports(referencedPackages, implementedPackages, exportedPackages).values());
    }
}
